gh-84649: Make TimedRotatingFileHandler use CTIME instead of MTIME

martonivan opened the pull request:
The TimedRotatingFileHandler previously used MTIME attribute of the log
file to detect whether it has to be rotate yet or not. In cases when the
file is changed within the rotation period the MTIME is also updated
to the current time and the rotation never happens. It's more
appropriate to check the file creation time (CTIME) instead.

<!--
Thanks for your contribution!
Please read this comment in its entirety. It's quite important.

# Pull Request title

It should be in the following format:

```
bpo-NNNN: Summary of the changes made
```

Where: bpo-NNNN refers to the issue number in the https://bugs.python.org.

Most PRs will require an issue number. Trivial changes, like fixing a typo, do not need an issue.

# Backport Pull Request title

If this is a backport PR (PR made against branches other than `master`),
please ensure that the PR title is in the following format:

```
[X.Y] <title from the original PR> (GH-NNNN)
```

Where: [X.Y] is the branch name, e.g. [3.6].

GH-NNNN refers to the PR number from `master`.

-->


<!-- issue-number: [bpo-40469](https://bugs.python.org/issue40469) -->
https://bugs.python.org/issue40469
<!-- /issue-number -->


<!-- gh-issue-number: gh-84649 -->
* Issue: gh-84649
<!-- /gh-issue-number -->

jacobtylerwalls commented on Misc/NEWS.d/next/Library/2021-02-26-13-17-57.bpo-40469.yJHeQg.rst:
```diff
@@ -0,0 +1,6 @@
+A bug has been fixed that made the ``TimedRotatingFileHandler`` use the
+MTIME attribute of the configured log file to to detect whether it has to be
+rotated yet or not. In cases when the file was changed within the rotatation
```
minor typo: rotation

martonivan replied:
Thanks for spotting this, it's fixed in the next commit. (I assumed that ammending the fix into the previous commit is not preferred.)

MaxwellDupre reviewed (dismissed):
Logging tests:
Ran 203 tests in 20.755s
OK (skipped=1) windoz
Look ok.

serhiy-storchaka reviewed (approved):
LGTM. I was just about to create the same issue.

martonivan committed:
gh-84649: Make TimedRotatingFileHandler use CTIME instead of MTIME

The TimedRotatingFileHandler previously only used st_mtime attribute of the
log file to detect whether it has to be rotate yet or not. In cases when the
file is changed within the rotatation period the st_mtime is also updated
to the current time and the rotation never happens.

It's more appropriate to check the file creation time (st_ctime) instead.
Whenever available, the more appropriate st_birthtime will be in use. (This
feature is available on FreeBSD, MacOS and Windows at the moment.) If
the st_mtime would be newer than st_ctime (e.g.: because the inode
related to the file has been changed without any file content
modification), then the earliest attribute will be used.

martonivan committed:
Add missing NEWS item about the change

serhiy-storchaka committed:
Merge branch 'main' into fix-issue-40469

serhiy-storchaka committed:
Re-wrap long lines.
